src/functor.jl

```diff
@@ -44,21 +44,73 @@ Equivalent to `functor(x)[1]`.
 """
 children(x) = functor(x)[1]
 
-function fmap1(f, x)
+function _default_walk(f, x)
   func, re = functor(x)
   re(map(f, func))
 end
 
-# See https://github.com/FluxML/Functors.jl/issues/2 for a discussion regarding the need for
-# cache.
-function fmap(f, x; exclude = isleaf, cache = IdDict())
+"""
+    fmap(f, x; exclude = isleaf, walk = Functors._default_walk)
+
+A structure and type preserving `map` that works for all [`functor`](@ref)s.
+
+By default, traveres `x` recursively using [`functor`](@ref)
+and transforms every leaf node identified by `exclude` with `f`.
+
+For advanced customization of the traversal behaviour, pass a custom `walk` function of the form `(f', xs) -> ...`.
+This function walks (maps) over `xs` calling the continuation `f'` to continue traversal.
+
+# Examples
+```jldoctest
+julia> struct Foo; x; y; end
+
+julia> @functor Foo
+
+julia> struct Bar; x; end
+
+julia> @functor Bar
+
+julia> m = Foo(Bar([1,2,3]), (4, 5));
+
+julia> fmap(x -> 2x, m)
+Foo(Bar([2, 4, 6]), (8, 10))
+
+julia> fmap(string, m)
+Foo(Bar("[1, 2, 3]"), ("4", "5"))
+
+julia> fmap(string, m, exclude = v -> v isa Bar)
+Foo("Bar([1, 2, 3])", (4, 5))
+```
```

Comment:
Do we want a custom walk example?

Reply:
I couldn't think of a succinct one that wasn't `fmapstructure`, so I just added a doctest for that instead. Are there any examples you can think of?

Reply:
Maybe just something that only walks the children that are arrays of arrays?

Reply:
Put in an absolutely uninspired example, but it's an example nonetheless. Happy to change it if something more illustrative comes up.